storage/innobase/log/log0online.cc

```diff
@@ -1061,17 +1149,50 @@ static void log_online_parse_redo_log(void) {
     mlog_id_t type;
     space_id_t space;
     page_no_t page_no;
+    bool is_single_rec =
+        (*(static_cast<const uchar *>(log_bmp_sys->parse_buf.ccurrent())) &
+         MLOG_SINGLE_REC_FLAG) != 0;
 
     const auto rec_parsed =
         log_bmp_sys->parse_buf.parse_next_record(&type, &space, &page_no);
+
     ut_ad(log_bmp_sys->parse_buf.get_current_lsn() <=
           log_bmp_sys->read_buf.get_current_lsn() + LOG_BLOCK_BOUNDARY_LSN_PAD);
-    if (rec_parsed && log_online_rec_has_page(type)) {
-      log_online_set_page_bit(space, page_no);
-      if (type == MLOG_INDEX_LOAD) {
-        const auto space_size = fil_space_get_size(space);
-        for (page_no_t i = 0; i < space_size; i++) {
-          log_online_set_page_bit(space, i);
+
+    if (rec_parsed) {
+      const auto last_parsed_lsn = log_bmp_sys->parse_buf.get_current_lsn();
+
+      if (log_bmp_sys->parse_buf.get_parse_start_lsn() == 0) {
+        // The first log record which should be checked is not yet known. Find
+        // it while parsing log records.
+        if (last_parsed_lsn >= log_bmp_sys->start_lsn &&
+            (is_single_rec || type == MLOG_DUMMY_RECORD ||
+             type == MLOG_MULTI_REC_END)) {
+          log_bmp_sys->parse_buf.set_parse_start_lsn(last_parsed_lsn);
+        }
+        continue;
+      }
+
+      if (log_bmp_sys->parse_buf.checkpoint_reached() &&
+          log_bmp_sys->parse_buf.get_parse_stop_lsn() == 0 &&
+          last_parsed_lsn >= log_bmp_sys->checkpoint_lsn &&
+          (is_single_rec || type == MLOG_DUMMY_RECORD ||
+           type == MLOG_MULTI_REC_END)) {
```

Comment:
This is for cases when first record group offset points before checkpoint_lsn. It may stop after checkpoint_lsn in the same block as well as continue searching in the next block.

Reply:
in such case, can't we simply use the checkpoint_lsn itself?

Reply:
Not sure I get your idea correctly. Can you please describe it in more detail.